### ISSUE TITLE
I get no-invalid-this error inside of a generated prototype method

### ISSUE THREAD

**Issue by unbornchikken** [eslint/eslint#8966]
* **ESLint Version:** 4.2
* **Node Version:** 8.1.1
* **npm Version:** 5.0.3

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

{
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "script"
  },
  "rules": {
    "no-invalid-this": 2
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js

BubuProvider.prototype.stop = Promise.method(function () {
	this.logInfo('Stopping BubuServer ...');
	this._httpServer.close();
});

```

**What did you expect to happen?**

Because the above is a generated, but still valid prototype method, using `this` in its context should not rise `no-invalid-this` assertion.

**What actually happened? Please include the actual, raw output from ESLint.**

* file: 'file:///c%3A/Dev/bimbo.js'
* severity: 'Error'
* message: 'Unexpected 'this'. (no-invalid-this)'
* at: '666,42'
* source: 'eslint'

I get the above error for positions of `this` keyword of the provided example above.

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, this is working as intended -- ESLint has no way of knowing that your function is a prototype method, because it isn't able to tell what `Promise.method` does with its argument.

**Comment by unbornchikken** [comment]
I believe the pattern that should accept `this` is the following:

Anything.prototype.AnyIdentifier = Anything(a-function-that-should-accept-this);
